src/flint/routers/api.py

```diff
@@ -43,12 +44,147 @@
+# https://developers.facebook.com/docs/graph-api/webhooks/getting-started#verification-requests
+def verify(request):
+    # Parse query params from the webhook verification request. Looking for hub.mode, hub.challenge, and hub.verify_token
+    mode = request.query_params.get("hub.mode")
+    token = request.query_params.get("hub.verify_token")
+    challenge = request.query_params.get("hub.challenge")
+    # Check if a token and mode were sent
+    if mode and token:
+        # Check the mode and token sent are correct
+        if mode == "subscribe" and token == verify_token:
+            # Respond with 200 OK and challenge token from the request
+            logger.info("WEBHOOK_VERIFIED")
+            return Response(challenge, status_code=200)
+        else:
+            # Responds with '403 Forbidden' if verify tokens do not match
+            logger.error("VERIFICATION_FAILED")
+            return Response(status_code=403)
+    else:
+        # Responds with '400 Bad Request' if verify tokens do not match
+        logger.error("MISSING_PARAMETER")
+        return Response(status_code=400)
+
+
+@api.get("/whatsapp_chat")
+async def whatsapp_chat(
+    request: Request,
+):
+    return verify(request)
+
+
+@api.post("/whatsapp_chat")
+async def whatsapp_chat_post(
+    request: Request,
+):
+    await handle_message(request)
+    return Response(status_code=200)
+
+
+def verified_body(body):
+    entry = body.get("entry")
+    if not entry:
+        return False
+    changes = entry[0].get("changes")
+    if not changes:
+        return False
+    value = changes[0].get("value")
+    if not value:
+        return False
+    messages = value.get("messages")
+    if not messages:
+        return False
+    return True
+
+
+# handle incoming webhook messages
+async def handle_message(request: Request):
+    # Parse Request body in json format
+    body = await request.json()
+    logger.info(f"request body: {body}")
+
+    try:
+        # info on WhatsApp text message payload:
+        # https://developers.facebook.com/docs/whatsapp/cloud-api/webhooks/payload-examples#text-messages
+        if body.get("object"):
+            if verified_body(body):
+                await handle_whatsapp_message(body)
+            return Response(status_code=200)
+        else:
+            # if the request is not a WhatsApp API event, return an error
+            return Response(status_code=404)
+    # catch all other errors and return an internal server error
+    except Exception as e:
+        logger.error(f"Error handling message: {e}", exc_info=True)
+        return Response(status_code=500)
+
+
+# handle WhatsApp messages of different type
+async def handle_whatsapp_message(body):
+    value = body["entry"][0]["changes"][0]["value"]
+    from_number = value["messages"][0]["from"]
+
+    # Get the user object
+    user = await sync_to_async(User.objects.prefetch_related("khojuser").filter)(khojuser__phone_number=from_number)
```

Comment:
Need to verify whether this `from_number` matches the format of the Twilio number -- if not, may need to tweak this to maintain consistency in user accounts.

Reply:
Edited to add -- Twilio number comes in with the `+`, while the WhatsApp cloud API number does not. Let's add the `+` to the phone number string